### PR TITLE
#17794 adding first draft for finding by name without taking in consi…

### PR DESCRIPTION
**Pull request by jdotcms**
Before this change, when you fire a workflow by name, it was using the available actions on listing and editing to get the actions, this was filtering by show when.
Consequently the user was not able to reach actions that were filtered by the show when logic.

Now, when an user sends a workflow name, the algorithm look for in the current step (if the content is in any), if can not found any action with the users name on the step, look for in the current scheme (all actions), otherwise tries to look for on all schemes actions.

This strategy might have a side effect, you can match an action with name that is not able in the step that the content is, but it is ok, since in the future this validation could be removed.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java [hunk 1]
```diff
@@ -178,7 +178,7 @@ private BulkActionView buildBulkActionView (final SearchResponse response,
         for (final Aggregation aggregation : aggregations.asList()) {
 
             if (aggregation instanceof StringTerms) {
-                StringTerms.class.cast(aggregation)
+                ((StringTerms) aggregation)
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: I didn;t change this. double check what is in master

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java [hunk 2]
```diff
@@ -1289,7 +1331,7 @@ private Role resolveRole(final String id) throws DotDataException {
 
         Role role = null;
         final String newid = id.substring
-                (id.indexOf("-") + 1, id.length());
+                (id.indexOf("-") + 1);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: I didn;t change this. double check what is in master

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java [hunk 3]
```diff
@@ -293,6 +293,20 @@ private BulkActionView buildBulkActionView (final Map<CountWorkflowStep, List<Wo
...
 
+    private Optional<WorkflowAction> findActionsOn (final String actionName, final List<WorkflowAction>  workflowActions) {
+
+        if (UtilMethods.isSet(workflowActions)) {
+
+            final Optional<WorkflowAction> foundAction  =
+                    workflowActions.stream()
+                            .filter(action -> action.getName().equalsIgnoreCase(actionName)).findFirst();
+
+            return foundAction;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: Codacy Issue found: [Consider simply returning the value vs storing it in local variable 'foundAction'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: Codacy Issue found: [String.indexOf(char) is faster than String.indexOf(String).](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: Codacy Issue found: [Consider simply returning the value vs storing it in local variable 'foundAction'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        //1 create a content type and associated to system workflow
+        final ContentType contentGenericType = new ContentTypeDataGen().workflowId(SystemWorkflowConstants.SYSTEM_WORKFLOW_ID)
+                .baseContentType(BaseContentType.CONTENT)
+                .field(new FieldDataGen().name("title").velocityVarName("title").next())
+                .field(new FieldDataGen().name("body").velocityVarName("body").next()).nextPersisted();
+
+        final String unicodeText = "Numéro de téléphone";
+        final ContentletDataGen contentletDataGen = new ContentletDataGen(contentGenericType.id());
+        Contentlet contentlet    = contentletDataGen.setProperty("title", "TestContent")
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: Codacy Issue found: [Local variable 'contentlet' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * Test {@link WorkflowHelper}
+ * @author jsanca
+ */
+public class TestWorkflowHelper extends IntegrationTestBase {
+
+    private static User user;
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: Codacy Issue found: [Consider simply returning the value vs storing it in local variable 'foundAction'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import com.dotcms.datagen.ContentTypeDataGen;
+import com.dotcms.datagen.ContentletDataGen;
+import com.dotcms.datagen.FieldDataGen;
+import com.dotcms.util.IntegrationTestInitService;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.portlets.contentlet.model.ContentletDependencies;
+import com.dotmarketing.portlets.contentlet.model.IndexPolicy;
+import com.dotmarketing.portlets.workflows.business.SystemWorkflowConstants;
+import com.dotmarketing.portlets.workflows.business.WorkflowAPI;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: Codacy Issue found: [Avoid unused imports such as 'com.dotmarketing.portlets.workflows.business.WorkflowAPI'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4801230)

**Review by jgambarios — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java (hunk 3 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: maybe rename to `getActionOnList`?

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java [hunk 7]
```diff
@@ -307,17 +321,45 @@ public String getActionIdByName(final String actionName,
...
+            }
+
+            foundAction = this.findActionsOn(actionName, workflowAPI.findActions(scheme, user));
+            if (foundAction.isPresent()) {
+
+                return foundAction.get().getId();
+            }
+        }
+
+        return null;
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: maybe this method should return a Optional

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    public static void prepare() throws Exception {
+        //Setting web app environment
+        IntegrationTestInitService.getInstance().init();
+        user = APILocator.getUserAPI().getSystemUser();
+    }
+
+    /**
+     * Test the getActionIdByName
+     */
+    @Test
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: improved doc in this test

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        final Contentlet contentletCheckin = APILocator.getWorkflowAPI().fireContentWorkflow(contentlet,
+                new ContentletDependencies.Builder().workflowActionId(saveActionId).modUser(user).indexPolicy(IndexPolicy.FORCE).build());
+
+        //4 get archive in a unpublish step
+        archiveActionId = workflowHelper.getActionIdByName("Archive", contentletCheckin, user);
+        Assert.assertEquals("The action returned should be Archive", SystemWorkflowConstants.WORKFLOW_ARCHIVE_ACTION_ID, archiveActionId);
+
+        //5 looking for non-existing Action
+        archiveActionId = workflowHelper.getActionIdByName("Non Existing Action", contentletCheckin, user);
+        Assert.assertNull("The action returned should be Archive", archiveActionId);
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: maybe this test could be split in several test

**Review by freddyucv — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java (hunk 8 above)
Comment (jdotcms) [reply dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: Done

File: dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java (hunk 4 above)
Comment (jdotcms) [reply dotCMS/src/integration-test/java/com/dotcms/workflow/helper/TestWorkflowHelper.java]: Done

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java (hunk 3 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: renamed to getActionIdOnList

File: dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java (hunk 7 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/workflow/helper/WorkflowHelper.java]: yes, but not than bad

**Pull request merged**